docs/guides/deno/byo-deep-research.mdx

```diff
@@ -0,0 +1,738 @@
+3. For each piece of new information:
+    a. If it's related to existing points, integrate it into the relevant paragraph.
+    b. If it's entirely new but relevant, add a new paragraph with a smooth transition.
+    c. If it's not relevant to the user topic, skip it.
+4. Ensure all additions are relevant to the user's topic.
+5. Verify that your final output differs from the input summary.
+</REQUIREMENTS>
+
+<FORMATTING>
+- Start directly with the updated summary, without preamble or titles. Do not use XML tags in the output.
+</FORMATTING>
+
+<Task>
+Think carefully about the provided Context first. Then generate a summary of the context to address the User Input.
+</Task>`
+```
+
+3. **Research Reflection** (`reflectionPrompt`): Analyzes current research to identify knowledge gaps
+
+```typescript title:prompts/reflect.ts
+export default (topics: string[]) => `
+You are an expert research assistant analyzing a summary about the following topics: 
+${topics.map((topic) => `- ${topic}`).join('\n')}.
+
+<GOAL>
+1. If the provided context is enough to answer the questions, do not generate a follow-up query.
+2. Identify knowledge gaps or areas that need deeper exploration
+3. Be careful not to generate follow-up queries that are not related to the topics.
+4. Generate a follow-up question that would help expand your understanding
+</GOAL>
+
+<REQUIREMENTS>
+Ensure the follow-up question is self-contained and includes necessary context for web search.
+</REQUIREMENTS>
+
+Provide only the follow-up query in your response, if there are no follow-up queries, respond with nothing.
+`
+```
+
+These prompts work together to create a research system that:
+
+- Generate search queries from topics
+- Find relevant content using the duckduckgo search API
+- Cleans and converts content to simple markdown
+- Summarizes findings
+- Attempts to identify knowledge gaps
+- Continues research if required
+
+### 3. Topic-based Research Pipeline
+
+The system uses a topic-based approach to conduct research, with different message types for each stage of the process:
+
+```typescript title:services/api.ts
+type TopicMessageTypes = 'create_query' | 'query' | 'reflect' | 'summarize'
+
+interface ResearchTopicMessage<T extends TopicMessageTypes> {
+  type: T
+  topics: string[]
+  summaries: string[]
+  remainingIterations: number
+}
+
+interface CreateQueryTopicMessage extends ResearchTopicMessage<'create_query'> {
+  type: 'create_query'
+  date: string
+  originalTopic: string
+}
+
+interface PerformQueryTopicMessage extends ResearchTopicMessage<'query'> {
+  type: 'query'
+  query: {
+    query: string
+    rationale: string
+  }
+}
+
+interface SummarizeTopicMessage extends ResearchTopicMessage<'summarize'> {
+  type: 'summarize'
+  content: string
+}
+
+interface ReflectTopicMessage extends ResearchTopicMessage<'reflect'> {
+  type: 'reflect'
+  content: string
+}
+```
+
+### 4. Content Processing
+
+The system includes HTML cleaning and markdown conversion:
```

Comment:
Maybe add additional information about the packages you're using and link to their website https://cheerio.js.org/. Otherwise I think the code is well documented.

Reply:
done